onecodex/cli.py

```diff
@@ -152,20 +156,70 @@ def samples(ctx, samples):
-@click.argument('files', nargs=-1, required=False,
-                type=click.Path(exists=True))
+@click.option('--max-threads', default=4,
+              help=OPTION_HELP['max_threads'], metavar='<int:threads>')
+@click.argument('files', nargs=-1, required=False, type=click.Path(exists=True))
+@click.option('--clean', is_flag=True, help=OPTION_HELP['clean'], default=False)
+@click.option('--interleave/--no-interleave', is_flag=True, help=OPTION_HELP['interleave'],
+              default=True)
+@click.option('--prompt/--no-prompt', is_flag=True, help=OPTION_HELP['prompt'], default=False)
 @click.pass_context
-def upload(ctx, files, no_threads, max_threads):
+def upload(ctx, files, no_threads, max_threads, clean, interleave, prompt):
     """Upload a FASTA or FASTQ (optionally gzip'd) to One Codex"""
-    if not no_threads:
-        ctx.obj['API'].Samples.upload(files, threads=max_threads)
-    else:
-        ctx.obj['API'].Samples.upload(files, threads=1)
+    if len(files) == 0:
+        print(ctx.get_help())
+        return
+
+    if interleave:
+        # "intelligently" find paired files and tuple them
+        paired_files = []
+        single_files = set(files)
+        for filename in files:
+            # convert "read 1" filenames into "read 2" and check that they exist; if they do
+            # upload the files as a pair, autointerleaving them
+            pair = re.sub('[._][Rr]1[._]', lambda x: x.group().replace('1', '2'), filename)
+            # we don't necessary need the R2 to have been passed in; we infer it anyways
+            if pair != filename and os.path.exists(pair):
+                if prompt:
+                    answer = click.confirm(
+                        'It appears {} and {} are paired. Upload an interleaved sample?'.format(
+                            os.path.basename(filename), os.path.basename(pair)
+                        ),
+                        default='Y'
+                    )
+                    if not answer:
+                        continue
+                elif pair not in single_files:
+                    # if we're not prompting, don't automatically pull in files not in the users
+                    # list
+                    continue
+
+                paired_files.append((filename, pair))
+                if pair in single_files:
+                    single_files.remove(pair)
+                single_files.remove(filename)
+        files = paired_files + list(single_files)
+
+    if not clean:
+        warnings.filterwarnings('error', category=ValidationWarning)
+
+    try:
+        # do the uploading
+        if not no_threads:
+            ctx.obj['API'].Samples.upload(files, threads=max_threads)
+        else:
+            ctx.obj['API'].Samples.upload(files, threads=1)
+    except ValidationWarning as e:
+        sys.stderr.write('\nERROR: {}. {}'.format(
+            e, 'Running with the --clean flag will suppress this error.'
+        ))
+    except ValidationError as e:
```

Comment:
This is the much improved version. The remaining gross part is we have to do the `sys.stderr.write` to get the output formatted properly newline-wise (so we can't use `click`s notifications)